students/Tianx/Lesson3/mailroom.py

```diff
@@ -0,0 +1,151 @@
+    else:
+        donor_list[name].insert(0, donation) # If donor name exists, insert the donation amount rather than updating the dictionary
+
+
+def send_thank_you():
+    """Sending a thank you email using user input.
+
+           Args:
+               None
+
+           Returns:
+               None
+
+           """
+
+    # Ask user for a donor's name or to display current list of donors, then ask for donation amount
+    print()
+    while True:
+        donor_name = input('======= The Thank you Menu: =======\n'
+                           "Enter 'list' for to see the list of donors\n"
+                           "or enter full name of donor \n"
+                           "Enter 'exit' to return to the main menu >")
+        if donor_name not in dict_of_donors and donor_name != "list" and donor_name != "exit":
+            print('\"{}\" is not a current donor, adding to the list...'.format(donor_name))
+            response = input('Please enter a donation amount for ' + donor_name + ' >')
+            donation_amount = float(response) # Convert the amount into a number
+            adding_donor_info(donor_name, donation_amount, dict_of_donors)
+            create_email(donor_name, donation_amount)
+        elif donor_name == "list": #If the user types list show them a list of the donor names and re-prompt.
+            show_donor_list(dict_of_donors)
+        elif donor_name == "exit": #If the user types exist return to main menu.
+            break
+        elif donor_name in dict_of_donors: #If user enters exsiting donor
+            response = input('Please enter a donation amount for ' + donor_name + ' >')
+            donation_amount = float(response)
+            adding_donor_info(donor_name, donation_amount, dict_of_donors)
+            create_email(donor_name, donation_amount)
+            break
+
+
+def create_email(name, donation):
+    print('=======Email Template=======')
+    print(f'Dear {name},\n\nThank you for your generousity, your donation of ${donation:.2f} will be put to good use.\n\n'
+      'Warm regards,\nMailroom Staff')
+
+
+def create_report():
+    """Formating a report.
+
+           Args:
+               None
+
+           Returns:
+               None
+
+           """
+    print('Donor Name                | Total Given | Num Gifts | Average Gift')
+    print('------------------------------------------------------------------')
+    for name in dict_of_donors:
+        total_given = sum(dict_of_donors[name])
+        number_gifts = len(dict_of_donors[name])
+        avg = total_given / number_gifts
+        print(f'{name:26} ${total_given:>11.2f} {number_gifts:>11.0f}  ${avg:>12.2f}')
+    print('------------------------------------------------------------------')
+
+
+def quit():
+    print("Exiting the menu now")
+    return "Exit Menu"
+
+
+main_dispatch = {"1": send_thank_you,
+                 '2': create_report,
+                 '3': quit,
+                 }
+
+if __name__ == '__main__':
```

Comment:
Code works and looks great!  Just missing the sorting function for the report.  Donors should be sorted by total amount given, most to least.